### PR TITLE
bugfix for `PathSimplifier`

### LINKED ISSUE #17914: `PathSimplifier` fails to ignore `CLOSEPOLY` vertices
### Bug report

**Bug summary**

This bug does not seem to have surfaced anywhere yet, but `Path`'s with `NaN` vertices at `CLOSEPOLY` can come through `PathNanRemover` in a "correct" state but then `PathSimplifier` can sometimes break it:

**Code for reproduction**/**Actual Outcome**

```python
>>> p = Path([(0, 0), (1, 0), (1, 1), (np.nan, np.nan)],
         [Path.MOVETO, Path.LINETO, Path.LINETO, Path.CLOSEPOLY])
>>> # correctly ignores CLOSEPOLY even though it's NaN
>>> p.cleaned(remove_nans=True)
Path(array([[ 0.,  0.],
       [ 1.,  0.],
       [ 1.,  1.],
       [nan, nan],
       [ 0.,  0.]]), array([ 1,  2,  2, 79,  0], dtype=uint8))
>>> # but now for some reason these NaN's are used to populate LINETO's
>>> p.cleaned(remove_nans=True, simplify=True)
Path(array([[ 0.,  0.],
       [ 1.,  0.],
       [ 1.,  1.],
       [nan, nan],
       [nan, nan],
       [ 0.,  0.]]), array([1, 2, 2, 2, 2, 0], dtype=uint8))
```

**Expected outcome**

The values of the vertices in a `CLOSEPOLY` should always be ignored, in favor of the most recent `MOVETO`'s vertex values:

```python
>>> p.cleaned(remove_nans=True, simplify=True)
Path(array([[ 0.,  0.],
       [ 1.,  0.],
       [ 1.,  1.],
       [0, 0],
       [ 0.,  0.]]), array([1, 2, 2, 2, 0], dtype=uint8))
```

**Matplotlib version**

  * Operating system:
  * Matplotlib version: 
  * Matplotlib backend (`print(matplotlib.get_backend())`):
  * Python version:
  * Jupyter version (if applicable):
  * Other libraries:

### PR DESCRIPTION
**Pull request by r3kste**
## PR summary

closes #17914

`PathSimplifier` should ignore vertices corresponding to `Path.CLOSEPOLY`, and rather replace them with the most recent `Path.MOVETO` vertex.

### Code for reproducing the bug / Actual outcome

```
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.path import Path
from matplotlib.patches import PathPatch

p = Path([(0, 0), (1, 0), (1, 1), (np.nan, np.nan)],
         [Path.MOVETO, Path.LINETO, Path.LINETO, Path.CLOSEPOLY])

# Remove this to see expected output
p = p.cleaned(simplify=True)

fig, ax = plt.subplots()
patch = PathPatch(p, facecolor="none", edgecolor="blue")
ax.add_patch(patch)
ax.set_xlim(-1, 2)
ax.set_ylim(-1, 2)

plt.show()
```

#### Expected Outcome
The Path should be closed.

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by r3kste** [comment]
As additional information, if the path had finite values instead on `np.nan` like:

```
p = Path([(0, 0), (1, 0), (1, 1), (0.5, 1.5)],
         [Path.MOVETO, Path.LINETO, Path.LINETO, Path.CLOSEPOLY])
```

#### Expected Outcome
The same as above.



#### Actual Outcome

**Comment by tacaswell** [comment]
Thank you for your work on this, but I do not this is the right solution.  It does make it look better, but simplify is still not handling the `CLOSEPOLY` correctly and is adding multiple LINETOs instead


```
In [2]: p.cleaned(simplify=True).codes
Out[2]: array([1, 2, 2, 2, 2, 0], dtype=uint8)

In [3]: p.codes
Out[3]: array([ 1,  2,  2, 79], dtype=uint8)


```

**Review by tacaswell — CHANGES_REQUESTED** [review]
I do not think that this is addressing the actual underlying bug.

**Comment by r3kste** [comment]
> but simplify is still not handling the `CLOSEPOLY` correctly and is adding multiple LINETOs instead

I think that this is not an issue with this handling of `CLOSEPOLY`, but rather that `PathSimplifier` repeats the last vertex twice, even if it was `Path.LINETO`.

```
>>> p         
Path(array([[0. , 0. ],
       [1. , 0. ],
       [1. , 1. ],
       [0.5, 1.5]]), array([1, 2, 2, 2], dtype=uint8))
>>> p.cleaned(simplify=True)
Path(array([[0. , 0. ],
       [1. , 0. ],
       [1. , 1. ],
       [0.5, 1.5],
       [0.5, 1.5],
       [0. , 0. ]]), array([1, 2, 2, 2, 2, 0], dtype=uint8))
```

> I do not think that this is addressing the actual underlying bug.

Do you mean that it should be handling `nan` values separately (as mentioned in the Issue discussion)?

**Comment by r3kste** [comment]
As I mentioned earlier, `PathSimplifier` adds the last vertex twice. It seems like this `queue_push` is what adds the vertex a second time.

```
--- a/src/path_converters.h
+++ b/src/path_converters.h
@@ -907,9 +907,6 @@ class PathSimplifier : protected EmbeddedQueue<9>
                 }
                 m_moveto = false;
             }
-            queue_push((m_moveto || m_after_moveto) ? agg::path_cmd_move_to : agg::path_cmd_line_to,
-                       m_lastx,
-                       m_lasty);
             m_moveto = false;
             queue_push(agg::path_cmd_stop, 0.0, 0.0);
         }
```

Removing these lines produces the "intended" behavior of `PathSimplifier`
```
>>> p
Path(array([[0., 0.],
       [1., 0.],
       [1., 1.]]), array([1, 2, 2], dtype=uint8))
>>> p.cleaned(simplify=True)
Path(array([[0., 0.],
       [1., 0.],
       [1., 1.],
       [0., 0.]]), array([1, 2, 2, 0], dtype=uint8))
```

However, I am not sure what would break if these lines are removed. Moreover, I think that it is better to address this issue in a different PR, because I feel that the underlying issue in #17914 was that `PathSimplifier` did not have logic for considering `CLOSEPOLY` vertices.

And for special casing of `nan` vertices, I think that this logic is already present in `PathNanRemover`, so I think including that in `PathSimplifier` might be unnecessary.
 
I would like to hear your thoughts on this @tacaswell

File: lib/matplotlib/tests/test_path.py [hunk 1]
```diff
@@ -541,3 +541,18 @@ def test_cleanup_closepoly():
         cleaned = p.cleaned(remove_nans=True)
         assert len(cleaned) == 1
         assert cleaned.codes[0] == Path.STOP
+
+
+def test_simplify_closepoly():
+    # The values of the vertices in a CLOSEPOLY should always be ignored,
+    # in favor of the most recent MOVETO's vertex values
+    path_ref = Path([(0, 0), (1, 0), (1, 1), (40, 50)],
```
Comment (greglucas) [inline lib/matplotlib/tests/test_path.py:549]: A few quick suggestions for enhancing this test:
* Use a different starting location from (0, 0) since that is inserted with the STOP command.
* It would be good to assert against the expected path directly rather than both of these using `simplify=True`
* Should we also see how `simplify=True` interacts with `remove_nans=True`? (see test above this)
* Is there any issue if the first MOVETO is `(nan, nan)` and then we insert it here?
* Could you test this with a compound path with two CLOSEPOLYs? (making sure the correct MOVETO gets inserted for each subpath)

**Review by greglucas — COMMENTED** [review]
This seems reasonable to me, but I think there could be a fair number of edge cases here that we could try to hit in the test to make it a bit more robust.

The extra LINETOs being added was noticed in the original issue: https://github.com/matplotlib/matplotlib/issues/17914#issuecomment-1671351396

Did you try running the test suite with your proposed change to remove that extra LINETO to see if it breaks anything?

File: lib/matplotlib/tests/test_path.py (hunk 1 above)
Comment (r3kste) [reply lib/matplotlib/tests/test_path.py:549]: > Should we also see how `simplify=True` interacts with `remove_nans=True`? (see test above this)

I am not sure what you meant by this. It would be helpful if you could elaborate a bit more.

> Is there any issue if the first MOVETO is `(nan, nan)` and then we insert it here?

Having the first MOVETO as `(nan, nan)` causes it to never initialize the initial vertex, and it ends up closing the polygon to (0, 0) which is wrong.

I think that in order to fix this, it could have a `m_has_init` so that when a CLOSEPOLY is encountered, it will be able to verify that the path has a finite initial point.

```
@@ -773,11 +780,11 @@ class PathSimplifier : protected EmbeddedQueue<9>
             }
             m_after_moveto = false;
 
-            if(agg::is_close(cmd)) {
+            if(agg::is_close(cmd) && m_has_init) {
                 /* If we have to close the polygon, replace
                    the vertex with the starting vertex */
-                *x = m_last_startx;
-                *y = m_last_starty;
+                *x = m_initX;
+                *y = m_initY;
             }
```

Refactoring `m_last_startx` to `m_initX` might be better, as that seems to be the variable name in other path converters (such as `NanRemover`)

**Comment by r3kste** [comment]
Thanks for the response.

> This seems reasonable to me, but I think there could be a fair number of edge cases here that we could try to hit in the test to make it a bit more robust.

I agree. I have modified the tests according to your suggestions.

> Did you try running the test suite with your proposed change to remove that extra LINETO to see if it breaks anything?

A significant number of tests in `test_simplification.py` fails. Most of these tests check for the size/number of vertices in the simplified path. These tests could be made to pass by fixing the off by one errors.

The only test that was not off by one was `test_start_with_moveto()`, which checks that the path starts with a `Path.MOVETO`. To account for this, I think a valid solution would be:

```
@@ -907,9 +907,10 @@ class PathSimplifier : protected EmbeddedQueue<9>
                 }
                 m_moveto = false;
             }
-            queue_push((m_moveto || m_after_moveto) ? agg::path_cmd_move_to : agg::path_cmd_line_to,
-                       m_lastx,
-                       m_lasty);
+
+            if (m_moveto || m_after_moveto) {
+                queue_push(agg::path_cmd_move_to, m_lastx, m_lasty);
+            }
             m_moveto = false;
             queue_push(agg::path_cmd_stop, 0.0, 0.0);
         }
```

This would guarantee that there is at least one vertex in the path, which would be a `MOVETO` to (0, 0).

Nevertheless, I couldn't find any tests where removing that extra LINETO produces a logical error.

File: lib/matplotlib/tests/test_path.py (hunk 1 above)
Comment (greglucas) [reply lib/matplotlib/tests/test_path.py]: >> Should we also see how simplify=True interacts with remove_nans=True? (see test above this)

> I am not sure what you meant by this. It would be helpful if you could elaborate a bit more.

I wanted to make sure that we weren't doing too much simplification if you pass `simplify=True` to a test similar to the one above this with only nan's. Making sure you still get back some valid path, which I think your other comment with assuring there is at least a single MOVETO does. So nothing to do here now.

File: lib/matplotlib/tests/test_path.py [hunk 2]
```diff
@@ -541,3 +541,35 @@ def test_cleanup_closepoly():
...
+        assert_array_equal(expected_path.vertices, simplified_path.vertices)
+        assert_array_equal(expected_path.codes, simplified_path.codes)
+
+    # test that a compound path also works
+    path = Path([(1, 1), (2, 1), (2, 2), (np.nan, np.nan),
+                 (-1, 0), (-2, 0), (-2, 1), (np.nan, np.nan)],
+                [Path.MOVETO, Path.LINETO, Path.LINETO, Path.CLOSEPOLY,
+                 Path.MOVETO, Path.LINETO, Path.LINETO, Path.CLOSEPOLY])
+    expected_path = Path([(1, 1), (2, 1), (2, 2), (1, 1),
+                          (-1, 0), (-2, 0), (-2, 1), (-1, 0), (-1, 0), (0, 0)],
```
Comment (greglucas) [inline lib/matplotlib/tests/test_path.py:568]: This is interesting that only the final CLOSEPOLY simplification gets the extra LINETO section.

I personally think it'd be nice to apply your fixes here and not assert against the extra LINETO since you've gone through the work of investigating them and your fix looks good to me at first glance.

Comment (greglucas) [reply lib/matplotlib/tests/test_path.py:568]: > However, what about the failing pytests in path_simplification.py
as mentioned in https://github.com/matplotlib/matplotlib/pull/28478#issuecomment-2241522679. Should I also modify them so as to fix the off by one errors?

I forgot about needing to update those as well. Maybe it is best to just leave the assertion with the extra expected LINETO for now and then update all the off-by-one issues in a follow-up PR.

**Review by greglucas — APPROVED** [review]
@tacaswell it would be good to get your eyes on this again since you're currently requesting changes.

**Comment by r3kste** [comment]
Thanks for reviewing the changes. 

I was thinking that moving the test from `test_path.py` to `test_simplification.py` would make more sense. I also have to refactor `m_last_startx` to `m_initX`.

These are small changes, so can I go ahead with them?

File: src/path_converters.h [hunk 3]
```diff
@@ -768,6 +784,19 @@ class PathSimplifier : protected EmbeddedQueue<9>
...
+            if(agg::is_close(cmd)) {
+                if (m_has_init) {
+                    /* If we have a valid initial vertex, then
+                       replace the current vertex with the initial vertex */
+                    *x = m_initX;
+                    *y = m_initY;
+                } else {
+                    /* If we don't have a valid initial vertex, then
+                       we can't close the path, so we skip the vertex */
+                    continue;
```
Comment (greglucas) [inline src/path_converters.h:796]: Could you add a test for this case as well?

What is the expected result if the second MOVETO of a compound path is `(nan, nan)`? Here it looks like we are still adding the inbetween LINETOs and just not starting/closing the path. Does it therefore merge into the first path in this case?

Comment (r3kste) [reply src/path_converters.h:796]: For a compound path such as:

```
p = Path([(1, 0), (1, -1), (2, -1),
          (np.nan, np.nan), (-1, -1), (-2, 1), (-1, 1),
          (2, 2), (0, -1)],
         [Path.MOVETO, Path.LINETO, Path.LINETO,
          Path.MOVETO, Path.LINETO, Path.LINETO, Path.LINETO,
          Path.CLOSEPOLY, Path.LINETO])
```

Here the second MOVETO is `(nan, nan)`. I plotted this path without simplification to see what the intended plot should be like. 



So it seems like, when plotting, if CLOSEPOLY is encountered without having a valid MOVETO vertex, it plots the point corresponding to the CLOSEPOLY (in this case, it is point 8 at `(2, 2)`) but doesn't close the polygon. After this, the LINETO to `(0, -1)` is continuous with the previously existing path.

> Here it looks like we are still adding the inbetween LINETOs and just not starting/closing the path. Does it therefore merge into the first path in this case?

So yes, you are correct. 

Therefore, the simplified Path would be:



```
>>> p.cleaned(simplify=True)
Path(array([[ 1.,  0.],
       [ 1., -1.],
       [ 2., -1.],
       [nan, nan],
       [-1., -1.],
       [-2.,  1.],
       [-1.,  1.],
       [ 0., -1.],
       [ 0., -1.],
       [ 0.,  0.]]), array([1, 2, 2, 1, 2, 2, 2, 2, 2, 0], dtype=uint8))
```

> Could you add a test for this case as well?

I think that the compound path discussed above will be good for the test.

**Comment by greglucas** [comment]
@tacaswell pinging you for a review on this since you're blocking. We are hitting some of these edge cases in Cartopy I think where we are currently working around them manually ourselves.

@r3kste do you have a separate update/PR with the removal of the extra LINETO you mentioned?

**Comment by r3kste** [comment]
> @r3kste do you have a separate update/PR with the removal of the extra LINETO you mentioned?

I haven't yet opened a separate PR for the removal of the extra LINETO, but I would like to do so.
Should I go ahead by opening one?

**Comment by greglucas** [comment]
> > @r3kste do you have a separate update/PR with the removal of the extra LINETO you mentioned?

> I haven't yet opened a separate PR for the removal of the extra LINETO, but I would like to do so.
Should I go ahead with opening one?

Yes, please feel free to open a PR with that update, I just wanted to make sure I hadn't missed it coming in.

**Review by tacaswell — APPROVED** [review]
I do not remember the previous code, but this seems correct to me now.

**Comment by greglucas** [comment]
Thanks, @r3kste and congratulations on your first merged Matplotlib PR 🎉  Hope to hear from you again!
